prime-dotnet-webapi/Services/AuthorizedUserService.cs

```diff
@@ -0,0 +1,133 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Security.Claims;
+using System.Threading.Tasks;
+using AutoMapper;
+using AutoMapper.QueryableExtensions;
+using Microsoft.AspNetCore.Http;
+using Microsoft.EntityFrameworkCore;
+using Prime.Extensions;
+using Prime.Models;
+using Prime.ViewModels.Parties;
+
+namespace Prime.Services
+{
+    public class AuthorizedUserService : BaseService, IAuthorizedUserService
+    {
+        private readonly IMapper _mapper;
+        private readonly IPartyService _partyService;
+
+        public AuthorizedUserService(
+            ApiDbContext context,
+            IMapper mapper,
+            IHttpContextAccessor httpContext,
+            IPartyService partyService)
+            : base(context, httpContext)
+        {
+            _mapper = mapper;
+            _partyService = partyService;
+        }
+
+        public async Task<bool> AuthorizedUserExistsAsync(int authorizedUserId)
+        {
+            return await _context.AuthorizedUsers
+                .AsNoTracking()
+                .AnyAsync(au => au.Id == authorizedUserId);
+        }
+
+        public async Task<bool> AuthorizedUserExistsForUserIdAsync(Guid userId)
+        {
+            return await _partyService.PartyExistsForUserIdAsync(userId, PartyType.AuthorizedUser);
+        }
+
+        public async Task<AuthorizedUserViewModel> GetAuthorizedUserAsync(int authorizedUserId)
+        {
+            return await GetBaseAuthorizedUserQuery()
+                .Where(au => au.Id == authorizedUserId)
+                .ProjectTo<AuthorizedUserViewModel>(_mapper.ConfigurationProvider)
+                .SingleOrDefaultAsync();
+        }
+
+        public async Task<AuthorizedUserViewModel> GetAuthorizedUserForUserIdAsync(Guid userId)
+        {
+            return await GetBaseAuthorizedUserQuery()
+                .Where(au => au.Party.UserId == userId)
+                .ProjectTo<AuthorizedUserViewModel>(_mapper.ConfigurationProvider)
+                .SingleOrDefaultAsync();
+        }
+
+        public async Task<int> CreateOrUpdateAuthorizedUserAsync(AuthorizedUserChangeModel changeModel, ClaimsPrincipal user)
+        {
+            var authorizedUser = await GetBaseAuthorizedUserQuery()
+                .SingleOrDefaultAsync(au => au.Party.UserId == user.GetPrimeUserId());
+
+            if (authorizedUser == null)
+            {
+                authorizedUser = new AuthorizedUser
+                {
+                    Party = new Party
+                    {
+                        Addresses = new List<PartyAddress>()
+                    },
+                    Status = AccessStatusType.UnderReview
+                };
+                _context.AuthorizedUsers.Add(authorizedUser);
```

Comment:
While this is now more succinct, it still throws if the User exists in the DB as a Party but not as an Authorized user.

Reply:
Yep, haven't updated it yet.